coprocessor deploy command

Nonnyjoe opened the pull request:
This PR closes: #231 

It runs a combination of the Carfile generation, publication and finally solidity contract deployment.
The current code implementation is able to find the respective folders for the solidity or coprocessor files provided the command is called from either, the root folder, the contract folder, or the coprocessor program folder.

Deployment is done using Viem and would require the user to pass in the name of the contract and also optional constructor arguments

Nonnyjoe committed:
feature: add coprocessor deploy command

tuler commented on apps/cli/src/commands/coprocessor.ts:
```diff
@@ -1,4 +1,6 @@
 import { Command } from "@commander-js/extra-typings";
+import { createDeployCommand } from "./coprocessor/deploy.js";
+import { createStartCommand } from "./coprocessor/start.js";
```
we removed start subcommand for now

tuler commented on apps/cli/src/compose/coprocessor/docker-compose-carize.yaml:
we use compose to execute service of the start command.
if you just need a one shot execution you could just run the docker container, no need for a compose

tuler commented on apps/cli/src/compose/coprocessor/docker-compose-carize.yaml:
```diff
@@ -0,0 +1,7 @@
+services:
+    carize:
+        image: ghcr.io/zippiehq/cartesi-carize:latest
```
we always use tagged versions, for stability

tuler replied:
Can the car generation be implemented in typescript and remove the need for an outside docker dependency?
I played a lot with CAR in typescript before, and it's usually straightforward.
The tradeoff is adding another npm dependency vs downloading another docker dependency.

https://www.npmjs.com/package/ipfs-car

Nonnyjoe replied:
Okay.... I'll check this out

tuler commented on apps/cli/src/commands/coprocessor/deploy.ts:
```diff
@@ -0,0 +1,490 @@
+};
+
+const publishCarFiles = async (cartesiDir: string): Promise<boolean> => {
+    const DEFAULT_OPERATOR_URL = "http://127.0.0.1:5001";
```
don't expect to have these ports exposed as the default.
they will be behind the `coprocessor start` proxy.

the method should receive an `options` object with these values.

tuler commented on apps/cli/src/commands/coprocessor/deploy.ts:
```diff
@@ -0,0 +1,490 @@
+    const DEFAULT_OPERATOR_URL = "http://127.0.0.1:5001";
+    const DEFAULT_SOLVER_URL = "http://127.0.0.1:3034";
+    const carFileName: string = "output.car";
+    const artifactsDir = path.resolve(cartesiDir, ".cartesi", "artifacts");
```
we avoid having `.cartesi` spread all over the code. That's why we created the `getContextPath` util method.

tuler commented on apps/cli/src/commands/coprocessor/deploy.ts:
```diff
@@ -0,0 +1,490 @@
+const publishCarFiles = async (cartesiDir: string): Promise<boolean> => {
+    const DEFAULT_OPERATOR_URL = "http://127.0.0.1:5001";
+    const DEFAULT_SOLVER_URL = "http://127.0.0.1:3034";
+    const carFileName: string = "output.car";
```
can we call this `image.car`?

Nonnyjoe replied:
Sure, it's possible, 
I'll work on that

tuler commented on apps/cli/src/commands/coprocessor/deploy.ts:
```diff
@@ -0,0 +1,490 @@
+    const carFileName: string = "output.car";
+    const artifactsDir = path.resolve(cartesiDir, ".cartesi", "artifacts");
+    const carFilePath = path.resolve(artifactsDir, carFileName);
+    const cidFilePath = path.resolve(artifactsDir, "output.cid");
```
who defines the spec of the CAR?
Are these names `output.cid`, `output.size` expected by some other component?

Nonnyjoe replied:
We can replace all reference of output with image.
I just need to confirm from @stskeeps that the operator is not requiring the file name to be output...

tuler commented:
I think deploying with viem that way you are doing is not good.
We are getting in the way of the developer with something that maybe we should not.
See my [latest comment about that](https://github.com/cartesi/cli/issues/156#issuecomment-2828603199).

tuler commented on apps/cli/src/commands/coprocessor/deploy.ts:
```diff
@@ -0,0 +1,490 @@
+                }
+
+                let { cartesiDir, foundryDir } = await findSubdirectories();
+                if (cartesiDir && foundryDir) {
```
Code structure is too complex.
Should be something like:

```typescript
const car = await pack({ ... });
await publish({ car ... });
await deploy({ ... });
```

exceptions raised by any step will stop execution and show the error.

Nonnyjoe commented:
@tuler 
In line with your latest reviews, I've moved the Carize generation process to a Carize.ts file and this required the addition of 2 important dependencies, which I did add to the project.

I also implemented other reviews, like renaming the generated carfiles from "output..." to "image..."

tuler commented on apps/cli/src/carize.ts:
```diff
@@ -0,0 +1,62 @@
+import path from "path";
+import { Writable } from "stream";
+
+export const createCarFromDirectory = async (
```
This is an util function.
In an util function don't call `console.log` or do `process.exit`.
Return whatever you need to return, that is needed outside.

Nonnyjoe committed:
feat(carize): added util function to generate car files

Nonnyjoe committed:
fixup! feature: add coprocessor deploy command

Nonnyjoe commented:
Had to rebase on top of the latest updates to `future/coprocessor-cli`, to avoid conflicts in the coprocessor.ts file and also to remove the previous commit that added the CI for carize generation

Nonnyjoe commented:
@tuler I made some updates to the Carize.ts in line with your reviews and also to make it fully compatible with the carfile unpacking process being used in the operator.

Nonnyjoe commented:
To test these modifications, you need to start the coprocessor devnet using the collowing steps:

- Clone the coprocessor repo with this command;
``` git clone  https://github.com/zippiehq/cartesi-coprocessor ```

- edit https://github.com/zippiehq/cartesi-coprocessor/blob/main/docker-compose-devnet.yaml#L29 this line to say:
```image: ghcr.io/zippiehq/cartesi-coprocessor-operator:pr-76```

- Run the commands:
``` 
docker compose -f docker-compose-devnet.yaml pull 

docker compose -f docker-compose-devnet.yaml up --wait -d
```

- Next you have to build this branch, create an alias, create a template, modify it, build the template and finally deploy your program using this a sample command similar to the below:

```
cartesi-dev coprocessor deploy MyContract --constructorArgs 0x8f86403A4DE0BB5791fa46B8e795C547942fE4Cf <machine hash>
```

**Note:** 
- Cartes-dev: Is an alias i used, you'll have to replace using yours.
- MyContract: is the name of the default contract in the template.
- "0x8f86403A4DE0BB5791fa46B8e795C547942fE4Cf" : Is the active task issuer address for the coprocessor ATM.

tuler commented:
Coprocessor initiative on hold.
